File: Java/DesignPatterns/Decorator_design_Pattern.java


interface Notification{
    void send(String message);
}

class SMSNotification implements Notification{
    @Override
    public void send(String message) {
        System.out.println("SMS is Sent: message" +message);
    }
}

class Whatsapp implements Notification{
    @Override
    public void send(String message) {
        System.out.println("whatsapp is Sent: message" +message);
    }
}

abstract class Decorator implements Notification{
    protected Notification wrapped_object;
    public Decorator(Notification wrapped_object){
        this.wrapped_object = wrapped_object;
    }
}

class RateLimiter extends Decorator{
    public RateLimiter(Notification wrapped_object) {
            super(wrapped_object);
    }
    
    @Override
    public void send(String message) {
        System.out.println("Logic of Rate limiting 1000 Lines");
        System.out.println("Logic of Rate limiting 1000 Lines");
        System.out.println("Logic of Rate limiting 1000 Lines");
        System.out.println("Logic of Rate limiting 1000 Lines");
        System.out.println("Logic of Rate limiting 1000 Lines");
        wrapped_object.send(message);
    }
}

class LoadBalancer extends Decorator{
    public LoadBalancer(Notification wrapped_object) {
        super(wrapped_object);
}

@Override
public void send(String message) {
        System.out.println("Logic Load balancer 1000 Lines");
        wrapped_object.send(message);
    }
}

class JSonFormatter extends Decorator{
    public JSonFormatter(Notification wrapped_object) {
        super(wrapped_object);
    }

    @Override
    public void send(String message) {
        System.out.println("Logic Json 1000 Lines");
        wrapped_object.send(message);
    }
}

class WhatsappFormtter extends Decorator{
    public WhatsappFormtter(Notification wrapped_object) {
        super(wrapped_object);
    }

    @Override
    public void send(String message) {
        System.out.println("whatsapp formetter");
        wrapped_object.send(message);
    }
}

public class Decorator_design_Pattern {
    public static void main(String[] args) {
        // Notification notification1 = new JSonFormatter(new RateLimiter(new LoadBalancer(new SMSNotification())));
        // notification1.send("pradeep"); 
        Notification notification2 = new WhatsappFormtter(new WhatsappFormtter(new RateLimiter((new Whatsapp()))));
        notification2.send("pradeep");          
    }    
}

File: JavaScript/DesignPatterns/Decorator_design_Pattern.js
class Notification {
    send(message) {
        throw new Error("Not implemented");
    }
}

class SMSNotification extends Notification {
    send(message) {
        console.log("SMS is Sent: message" + message);
    }
}

class Whatsapp extends Notification {
    send(message) {
        console.log("whatsapp is Sent: message" + message);
    }
}

class Decorator extends Notification {
    constructor(wrapped_object) {
        super();
        this.wrapped_object = wrapped_object;
    }
}

class RateLimiter extends Decorator {
    constructor(wrapped_object) {
        super(wrapped_object);
    }

    send(message) {
        console.log("Logic of Rate limiting 1000 Lines");
        console.log("Logic of Rate limiting 1000 Lines");
        console.log("Logic of Rate limiting 1000 Lines");
        console.log("Logic of Rate limiting 1000 Lines");
        console.log("Logic of Rate limiting 1000 Lines");
        this.wrapped_object.send(message);
    }
}

class LoadBalancer extends Decorator {
    constructor(wrapped_object) {
        super(wrapped_object);
    }

    send(message) {
        console.log("Logic Load balancer 1000 Lines");
        this.wrapped_object.send(message);
    }
}

class JSonFormatter extends Decorator {
    constructor(wrapped_object) {
        super(wrapped_object);
    }

    send(message) {
        console.log("Logic Json 1000 Lines");
        this.wrapped_object.send(message);
    }
}

class WhatsappFormtter extends Decorator {
    constructor(wrapped_object) {
        super(wrapped_object);
    }

    send(message) {
        console.log("whatsapp formetter");
        this.wrapped_object.send(message);
    }
}

class Decorator_design_Pattern {
    static main(args) {
        // Notification notification1 = new JSonFormatter(new RateLimiter(new LoadBalancer(new SMSNotification())));
        // notification1.send("pradeep");
        const notification2 = new WhatsappFormtter(new WhatsappFormtter(new RateLimiter((new Whatsapp()))));
        notification2.send("pradeep");
    }
}

Decorator_design_Pattern.main([]);
